### PR TITLE
push watch: use queue when pushing, wait on queue to empty upon ctrl-c

### LINKED ISSUE #53: Add signal handling to the daemon
In order to avoid having to call `cachix push` [twice](51), we can do the following:

* Add signal handling to the daemon mode
* Have a signal that, when received, finishes pushing and then exits
* Run the daemon in a job in parallel with `nix-build`
* Send that signal after `nix-build`

### PR DESCRIPTION
**Pull request by domenkozar**
This also contains some breaking changes:

- `Push` module now operates with `PushCache m r` that simplifies how many arguments need to be passed around
- watch mode doesn't upload `drv` files anymore

Fixes #53 #339

Example:

```
Watching /nix/store for new paths ...
compressing and pushing /nix/store/g6si4dzzxvv0vibrc1sa1qdirgxv9v1i-1609944078-1 (976.67 KiB)
compressing and pushing /nix/store/2pf7xljj84z5f0880bsgkbipndah39zp-empty-10-1MB (680.00 B)
compressing and pushing /nix/store/243apdjx4ddwk45kijmlc99kdnsy5d24-1609944078-7 (976.67 KiB)
compressing and pushing /nix/store/6x43igsqvg7g6j5rnpa0g4m6rbqm5nbk-1609944078-3 (976.67 KiB)
compressing and pushing /nix/store/baabkwgbrmjz9m672x3x57pdjp0ah9x5-1609944078-9 (976.67 KiB)
^CStopping producer of store paths and waiting for queue to empty ...
Waiting to finish: 5 pushing, 6 in queue
compressing and pushing /nix/store/c3q7dnd9zg32azz1jglyzdi8yi5yzi5v-1609944078-5 (976.67 KiB)
Waiting to finish: 5 pushing, 5 in queue
compressing and pushing /nix/store/h2x8ibnwvswnv16wry71fn6vajljqdqg-1609944078-10 (976.67 KiB)
Waiting to finish: 5 pushing, 4 in queue
compressing and pushing /nix/store/k859flmfrnppmf4r4w0x20sp7k0cn8lg-1609944078-6 (976.67 KiB)
Waiting to finish: 5 pushing, 3 in queue
compressing and pushing /nix/store/plnp0l8y4h9scl7zzvjr4i7h8fcyj1yb-1609944078-8 (976.67 KiB)
Waiting to finish: 5 pushing, 2 in queue
compressing and pushing /nix/store/pzakihpgprjvfs92z1hpdr4mgcsnhjza-1609944078-2 (976.67 KiB)
compressing and pushing /nix/store/yvw3may5yphfxfpi1dqfb5xx7w8zvd2z-1609944078-4 (976.67 KiB)
Waiting to finish: 5 pushing, 0 in queue
Waiting to finish: 5 pushing, 0 in queue
Waiting to finish: 4 pushing, 0 in queue
Waiting to finish: 1 pushing, 0 in queue
Waiting to finish: 1 pushing, 0 in queue
Waiting to finish: 1 pushing, 0 in queue
Waiting to finish: 1 pushing, 0 in queue
Waiting to finish: 1 pushing, 0 in queue
Waiting to finish: 1 pushing, 0 in queue
Waiting to finish: 1 pushing, 0 in queue
All done, exiting.
```

### CODE REVIEW

File: cachix/src/Cachix/Client/PushQueue.hs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  = WorkerState
+      { inQueue :: Queue,
+        inProgress :: TVar (S.Set StorePath)
+      }
+
+worker :: Push.PushCache IO () -> WorkerState -> IO ()
+worker pushCache workerState = forever $ do
+  storePath <- atomically $ TBQueue.readTBQueue $ inQueue workerState
+  inprogress <- readTVarIO $ inProgress workerState
+  -- TODO: there's a small window between checking membership and inserting new path
```
Comment (roberth) [inline cachix/src/Cachix/Client/PushQueue.hs:37]: Your first `atomically` can take from the queue, update `inProgress` and return `Maybe StorePath`.
Then `for` the maybe, you do the work and at then end update `inProgress` in the second transaction.

File: cachix/src/Cachix/Client/PushQueue.hs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  unless (S.member storePath inprogress) $ bracket_ (start storePath) (stop storePath) $
+    Push.pushSingleStorePath pushCache storePath
+  where
+    start storePath =
+      atomically $ modifyTVar' (inProgress workerState) (S.insert storePath)
+    stop storePath =
+      atomically $ modifyTVar' (inProgress workerState) (S.delete storePath)
+
+-- NOTE: producer is responsible for signaling SIGINT upon termination
+pushQueue :: Int -> (Queue -> IO (IO ())) -> Push.PushCache IO () -> IO ()
```
Comment (roberth) [inline cachix/src/Cachix/Client/PushQueue.hs:47]: Perhaps `withNewPushWorker`

File: cachix/src/Cachix/Client/PushQueue.hs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  stopProducerCallback
+  go
+  where
+    go = do
+      queueLength <- atomically $ TBQueue.lengthTBQueue $ inQueue workerState
+      inprogress <- readTVarIO $ inProgress workerState
+      if queueLength == 0 && S.null inprogress
+        then do
+          putText "All done, exiting."
+          uninterruptibleCancel consumer
```
Comment (roberth) [inline cachix/src/Cachix/Client/PushQueue.hs:69]: Potential race condition here. It's nicer to put an "EOF" on the queue, so the threads can exit by themselves and you just wait for the consumer async to complete here.

Comment (domenkozar) [reply cachix/src/Cachix/Client/PushQueue.hs:69]: What is the race condition that can happen?

Comment (roberth) [reply cachix/src/Cachix/Client/PushQueue.hs:69]: I guess this is more of a general advice than specific to cachix.

 - thread a creates thread b
 - thread b adds to queue
 - thread a terminates the worker
 - thread a assumes everything is done
 - thread b adds to queue

It probably doesn't apply here.

Comment (domenkozar) [reply cachix/src/Cachix/Client/PushQueue.hs:69]: On line 60 it terminates the producer, I should note that it should be a blocking action.

I tried the approach with EOF but it makes it more complicated to report number of items in queue.

**Comment by domenkozar** [comment]
Also need to implement batching for narinfos.

**Comment by domenkozar** [comment]
@roberth still need to figure out how to wait both queues to finish, otherwise query queue is there.

File: cachix/src/Cachix/Client/PushQueue.hs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  consumer <- async $ replicateConcurrently_ numWorkers $ worker pushCache workerState
+  void $ Signals.installHandler Signals.sigINT (Signals.CatchOnce (exitOnceQueueIsEmpty stopProducerCallback consumer workerState)) Nothing
+  waitEither_ consumer queryWorker
+
+queryLoop :: QueryWorkerState -> Queue -> Push.PushCache IO () -> IO ()
+queryLoop workerState pushQueue pushCache = do
+  storePaths <- atomically $ do
+    -- this blocks until item is available and doesn't remove it from the queue
+    _ <- TBQueue.peekTBQueue (queryQueue workerState)
+    TBQueue.flushTBQueue (queryQueue workerState)
```
Comment (roberth) [inline cachix/src/Cachix/Client/PushQueue.hs:72]: Equivalent to flush + `guard` not empty, fwiw

**Comment by roberth** [comment]
> @roberth still need to figure out how to wait both queues to finish, otherwise query queue is there.

You could probably introduce a `TVar Bool` to indicate whether you have in-flight queries and check that one as well in the same `atomically` as where you check that the worker queue is empty. And you should check the `queryQueue` itself there as well. That should do.

If you want a nicer design, you could consider tracking the status of the store paths themselves in `TVar`s that accompany the `StorePath`s. That way you can decouple the termination logic from the queues.

File: cachix/src/Cachix/Client/PushQueue.hs (hunk 4 above)
Comment (domenkozar) [reply cachix/src/Cachix/Client/PushQueue.hs:72]: Flush says it never retries, so it would then need handle the loop

**Comment by domenkozar** [comment]
@roberth alright this is ready - there's one problem though, sigint is still caught in the subprocess `nix-store --dump` so maybe time to revive #312

File: cachix/src/Cachix/Client/PushQueue.hs (hunk 4 above)
Comment (roberth) [reply cachix/src/Cachix/Client/PushQueue.hs]: The `guard` would retry it.

**Comment by roberth** [comment]
Sounds like `delegate_ctlc` in `System.Process`. http://hackage.haskell.org/package/process-1.6.10.0/docs/System-Process.html#g:4

**Comment by domenkozar** [comment]
That one doesn't help, but I found out that `create_group` makes it ignore signals. Probably not the best idea in case those processes hang?

**Comment by domenkozar** [comment]
Alright it should still work since removing stdin makes them crash. Not the best, but it works.

**Comment by roberth** [comment]
> That one doesn't help, but I found out that `create_group` makes it ignore signals. Probably not the best idea in case those processes hang?

If you want to protect against hanging you'd need a `timeout` and an extra exception handler because the one in `createProcess` only sends a SIGTERM iirc. I don't think you want to be paranoid like that though.

**Comment by roberth** [comment]
Although the `timeout` isn't necessary when your Ctrl+C causes the exception.

**Comment by domenkozar** [comment]
I've tested SIGINT and SIGTERM and it behaves as expected now.

**Comment by domenkozar** [comment]
Did more testing, it works! @roberth any objections for merging?

**Review by roberth — COMMENTED** [review]
Looks alright with the small naming improvements.

I might still prefer a custom solution for hercules-ci-agent, because in a multi-process situation I want to know when a path is done without shutting down the worker. So I'll probably use the low-level `uploadStorePath`/`pushSingleStorePath` and call the `narInfoBulk` endpoint myself as well.
I'd like to be able to reuse `PushQueue` but I feel like that's unlikely to be a good fit with support for other caches and the multi-process nature of the agent. Also Hercules CI already uploads store paths immediately so the immediate benefit is small. (more efficient bulk queries)

To summarize, I think this is a good solution for the `cachix` CLI and I can still use it on the agent. So no objections :)

**Comment by domenkozar** [comment]
> I might still prefer a custom solution for hercules-ci-agent, because in a multi-process situation I want to know when a path is done without shutting down the worker. So I'll probably use the low-level uploadStorePath/pushSingleStorePath and call the narInfoBulk endpoint myself as well.

If you come up with a way to generalize the current solution that seems acceptable to me. I also want to support #343 that will require changes to work in a daemon mode.

**Comment by roberth** [comment]
> > I might still prefer a custom solution for hercules-ci-agent, because in a multi-process situation I want to know when a path is done without shutting down the worker. So I'll probably use the low-level uploadStorePath/pushSingleStorePath and call the narInfoBulk endpoint myself as well.
> 
> If you come up with a way to generalize the current solution that seems acceptable to me. I also want to support #343 that will require changes to work in a daemon mode.

If the caller can provide a `(StorePath, TVar Bool)` instead of just a storepath and the code sets it to true when done, that should go a long way, but I'm not sure if that's worthwhile for hercules-ci-agent because it needs to generalize to other cache types besides cachix. I suppose that might be useful in for your #343 too. Idk

**Comment by domenkozar** [comment]
Other caches don't have a good way to query for closure in efficient way, but lets leave generalizations for when that's needed :)
